extensions/amp-analytics/0.1/test/test-crypto-impl.js

```diff
@@ -18,28 +18,58 @@ import {Crypto} from '../crypto-impl';
 
 describe('crypto-impl', () => {
 
-  let crypto;
+  function testSuite(descption, crypto) {
+    describe(descption, () => {
+      it('should hash "abc" in sha384', () => {
+        return crypto.sha384('abc').then(buffer => {
+          expect(buffer.length).to.equal(48);
+          expect(buffer[0]).to.equal(203);
+          expect(buffer[1]).to.equal(0);
+          expect(buffer[47]).to.equal(167);
+        });
+      });
+      
+      it('should hash "abc" in sha384Base64', () => {
+        return expect(crypto.sha384Base64('abc')).to.eventually.equal(
+            'ywB1P0WjXou1oD1pmsZQBycsMqsO3tFjGotgWkP_W-2AhgcroefMI1i67KE0yCWn');
+      });
 
-  beforeEach(() => {
-    crypto = new Crypto({});
-  });
+      it('should hash "foobar" in sha384Base64', () => {
+        return expect(crypto.sha384Base64('foobar')).to.eventually.equal(
+            'PJww2fZl501RXIQpYNSkUcg6ASX9Pec5LXs3IxrxDHLqWK7fzfiaV2W_kCr5Ps8G');
+      });
 
-  it('should hash "abc" in sha384', () => {
-    return crypto.sha384('abc').then(buffer => {
-      expect(buffer.length).to.equal(48);
-      expect(buffer[0]).to.equal(203);
-      expect(buffer[1]).to.equal(0);
-      expect(buffer[47]).to.equal(167);
+      it('should throw when input contains chars out of range [0,255]', () => {
+        expect(() => crypto.sha384('abc今')).to.throw();
+        expect(() => crypto.sha384Base64('abc今')).to.throw();
+      });
     });
-  });
+  }
 
-  it('should hash "abc" in sha384Base64', () => {
-    return expect(crypto.sha384Base64('abc')).to.eventually.equal(
-        'ywB1P0WjXou1oD1pmsZQBycsMqsO3tFjGotgWkP_W-2AhgcroefMI1i67KE0yCWn');
-  });
+  testSuite('with native crypto API', new Crypto(window));
+  testSuite('with crypto lib', new Crypto({}));
+  testSuite('with native crypto API rejects', new Crypto({
+    crypto: {
+      subtle: {
+        digest: () => Promise.reject('Operation not supported'),
+      },
+    },
+  }));
+  testSuite('with native crypto API throws', new Crypto({
+    crypto: {
+      subtle: {
+        digest: () => {
+          throw new Error();
+        },
+      },
+    },
+  }));
 
-  it('should hash "foobar" in sha384Base64', () => {
-    return expect(crypto.sha384Base64('foobar')).to.eventually.equal(
-        'PJww2fZl501RXIQpYNSkUcg6ASX9Pec5LXs3IxrxDHLqWK7fzfiaV2W_kCr5Ps8G');
+  it('native API result should exactly equal to crypto lib result', () => {
+    return Promise
```

Comment:
Can you make sure that native API is actually called? E.g. by counting calls into fallback?